org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java

```diff
@@ -0,0 +1,150 @@
+					if (project.exists()) {
+
+						// Check if it's sitting in the output dir, look for original one
+						if (project.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+							IPath relative = resource.getFullPath().removeFirstSegments(project.getOutputLocation().segmentCount());
+
+							for (IPackageFragmentRoot root : project.getPackageFragmentRoots()) {
+								if (root.getKind() == IPackageFragmentRoot.K_SOURCE) {
+									IPath fullPath = root.getCorrespondingResource().getProjectRelativePath().append(relative);
+									IFile newFile = resource.getProject().getFile(fullPath);
+									if (newFile.exists()) {
+										return new FileEditorInput(newFile);
+									}
+								}
+							}
+						}
+
+						// Check if it's sitting in one of the output folders set on the source folders
+						for (IClasspathEntry classpathEntry : project.getRawClasspath()) {
+							if (classpathEntry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+								if (classpathEntry.getOutputLocation() != null
+										&& classpathEntry.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+									IPath relative = resource.getFullPath()
+											.removeFirstSegments(classpathEntry.getOutputLocation().segmentCount());
+									for (IPackageFragmentRoot root : project.findPackageFragmentRoots(classpathEntry)) {
+										IPath fullPath = root.getCorrespondingResource().getProjectRelativePath().append(relative);
+										IFile newFile = resource.getProject().getFile(fullPath);
+										if (newFile.exists()) {
+											return new FileEditorInput(newFile);
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		} catch (JavaModelException e) {
+			if (e.isDoesNotExist()) {
+				LOG.error("Error finding the original source for the editor input", e);
+			}
+		}
+
+		return input;
+	}
+
+	public IEditorInput findOriginalSource(IEditorInput input) {
+		IFile resource = ResourceUtil.getFile(input);
+		if (resource == null) {
+			return input;
+		}
+
+		IEditorInput original = findOriginalSourceForOuputFolderCopy(input);
+		if (original != input) {
+			return original;
+		}
+
+		IEclipseTrace trace = traceInformation.getTraceToSource(resource);
+		if (trace == null) {
+			return input;
+		}
+
+		for (ILocationInEclipseResource candidate : trace.getAllAssociatedLocations()) {
+			if (languageInfo.equals(candidate.getLanguage())) {
+				IStorage storage = candidate.getPlatformResource();
+				if (storage != null) {
+					return EditorUtils.createEditorInput(storage);
+				}
+			}
+		}
+
+		return input;
+	}
+
+	/**
+	 * @return the {@code IPackageFragmentRoot} that contains the provided {@code element}
+	 * @throws NullPointerException
+	 *             if the provided {@code element} is null
+	 */
+	public IPackageFragmentRoot getPackageFragmentRoot(IJavaElement element) {
+		Preconditions.checkNotNull(element, "Couldn't find source package fragment root.");
+		return element instanceof IPackageFragmentRoot //
```

Comment:
The semantics of getAncestor do already imply the check of the element itself. Can be implemented as `return (IPackageFragmentRoot) element.getAncestor()`